### PR TITLE
Ephemeral storage requests/limits

### PR DESCRIPTION
**Pull request by Joel-hanson**
### Type of change

- Enhancement / new feature
- Documentation

### Description

What does this PR do?

1. This PR configures requests and limits for ephemeral storage in each product container for pods that utilise it.
2. If the following circumstances are fulfilled, it updates all container pods with ephemeral storage requests and limits:

  - Request - If the volume is either `emptyDir`, `configMap`, `downardAPI`, or `secret`, the ephemeral-storage request is set for the container pod resources in all the circumstances.
  - Limit - For all container pod resources, an ephemeral-storage limit has been set:

    1. If the container security context is non-ReadOnlyRootFilesystem and the `emptyDir` volume doesn't have a sizeLimit.
    2. If the container security context is non-ReadOnlyRootFilesystem and it is one of `configMap,` `downardAPI,` or`secret` then the ephemeral-storage limit is set.

3. Updated Pod template to have the configuration for ephemeral-storage request and ephemeral-storage limit.
4. The test for the above implementation.
5. The default value of the ephemeral storage is set as `1 Gi`

**_Note: The default values for the ephemeral storage is set to `1 Gi` because the teams with the potential need of more than 1GB of local ephemeral storage should consider different options such as persistent storage for its own PVC rather than relying on ephemeral storage._**

<details>
  <summary>Detailed explanation</summary>

### Overview

Ephemeral storage is any storage that is not guaranteed to be persisted (in Kubernetes, ephemeral storage on the node is provided by locally-attached writeable devices or, sometimes, by RAM), this is usually an emptyDir or a writable layer in the container.

Ephemeral storage, like all storage, has the option to declare limits and requests for the Kubernetes runtime. Products that do not declare ephemeral storage limits and requests can give an inaccurate picture to the Kubernetes runtime for what their storage requirements are (in terms of memory and cpu units), and this can lead to clusters overcommitting on data requests that cannot be fulfilled.

**This requires ephemeral-storage requests in all scenarios. For emptyDir - ephemeral-storage limits are needed when the container is non-ReadOnlyRootFilesystem and the emptyDir volume has no sizeLimit. For configMap, downardAPI, and secret ephemeral-storage limits are needed when the container is non-ReadOnlyRootFilesystem**

### Instructions

1. Products MUST declare ephemeral storage requests for each product container in the Pod definition
`spec.containers[].resources.request.ephemeral-storage`

2. Product MUST declare ephemeral storage limits for each product container in the Pod definition using the following guidance:

- All emptyDir volumes require a sizeLimit: `emptyDir.sizeLimit`
- If you use the `readOnlyRootFilesystem` security context, do not set ephemeral limits.
- If you do not use a read only root filesystem, set a limit of `500Mi` (for logs) plus whatever you expect to be reasonable usage of the container root filesystem, the emptyDirs, and the mounted configmaps/secrets.

</details>

<details>
  <summary>Examples</summary>

#### Passing Examples

**Example: EmptyDir**

_Pod.yaml (rendered from test default)_:

    apiVersion: v1
    kind: Pod
    metadata:
      name: emptyDirExample
    spec:
      containers:
      - image: docker.io/sample-panamax-nginx:latest
        name: nginx
        resources:
          limits:
            ephemeral-storage: 6Mi
          requests:
            ephemeral-storage: 2Mi
        volumeMounts:
        - mountPath: /cache
          name: anEmptyDir
      - image: docker.io/my-example-frontend-ui:latest
        name: ui
        resources:
          limits:
            ephemeral-storage: 5Gi
          requests:
            ephemeral-storage: 1Gi
        volumeMounts:
        - mountPath: /cache
          name: anEmptyDir
      volumes:
      - emptyDir: {}
        name: anEmptyDir

<hr>

**Example: EmptyDirReadOnlyNoESLimit**

_Pod.yaml (rendered from test default)_:

    apiVersion: v1
    kind: Pod
    metadata:
      name: emptyDirReadOnlyNoESLimit
    spec:
      containers:
      - image: docker.io/sample-panamax-nginx:latest
        name: nginx
        resources:
          requests:
            ephemeral-storage: 2Mi
        securityContext:
          readOnlyRootFilesystem: true
        volumeMounts:
        - mountPath: /cache
          name: anEmptyDir
      - image: docker.io/my-example-frontend-ui:latest
        name: ui
        resources:
          requests:
            ephemeral-storage: 1Gi
        securityContext:
          readOnlyRootFilesystem: true
        volumeMounts:
        - mountPath: /cache
          name: anEmptyDir
      volumes:
      - emptyDir:
          sizeLimit: 1Gi
        name: anEmptyDir

<hr>

**Example: No Local Ephemeral Storage Used**

_Pod.yaml (rendered from test default)_:

    apiVersion: v1
    kind: Pod
    metadata:
      name: noEphemeralStorage
    spec:
      containers:
      - image: docker.io/sample-panamax-nginx:latest
        name: nginx
      - image: docker.io/my-example-frontend-ui:latest
        name: ui

<hr>

**Example: Passing ConfigMap, DownwardAPI, and Secret Limit and Request on non-readOnly**

_Pod.yaml (rendered from test default)_:

    apiVersion: v1
    kind: Pod
    metadata:
      name: passingConfigMap
    spec:
      containers:
      - image: docker.io/sample-panamax-nginx:latest
        name: nginx
        resources:
          limits:
            ephemeral-storage: 6Mi
          requests:
            ephemeral-storage: 2Mi
            memory: 32Mi
        securityContext:
          readOnlyRootFilesystem: false
        volumeMounts:
        - mountPath: /cache
          name: aConfigMap
        - mountPath: /secret-location
          name: aSecret
        - mountPath: /api
          name: aDownwardAPI
      volumes:
      - configMap:
          items:
          - keys: jsonString
            path: data
          name: targetConfigMap
        name: aConfigMap
      - name: aSecret
        secret:
          secretName: targetSecret
      - downwardAPI:
          items:
          - path: memory_request
          - resourceFieldRef:
              containerName: nginx
              resource: requests.memory
        name: aDownwardAPI

<hr>

#### Failing Examples

**Example: Ephemeral Storage Used But Not Declared**

_Pod.yaml (rendered from test default)_:

    apiVersion: v1
    kind: Pod
    metadata:
      name: emptyDirExample
    spec:
      containers:
      - image: docker.io/sample-panamax-nginx:latest
        name: nginx
        resources: {}
        volumeMounts:
        - mountPath: /cache
          name: anEmptyDir
      - image: docker.io/my-example-frontend-ui:latest
        name: ui
        resources: {}
        volumeMounts:
        - mountPath: /cache
          name: anEmptyDir
      volumes:
      - emptyDir: {}
        name: anEmptyDir

Messages:
* `[ERROR] Pod.yaml: container "nginx" mounts local ephemeral storage of type emptyDir, but ephemeral storage limits and requests are not declared (EphemeralStorageLimitsAndRequestsDeclared)`
* `[ERROR] Pod.yaml: container "ui" mounts local ephemeral storage of type emptyDir, but ephemeral storage limits and requests are not declared (EphemeralStorageLimitsAndRequestsDeclared)`
<hr>

**Example: Ephemeral Storage Used But No Limits Declared**

_Pod.yaml (rendered from test default)_:

    apiVersion: v1
    kind: Pod
    metadata:
      name: emptyDirExample
    spec:
      containers:
      - image: docker.io/sample-panamax-nginx:latest
        name: nginx
        resources:
          requests:
            ephemeral-storage: 2Mi
        volumeMounts:
        - mountPath: /cache
          name: anEmptyDir
      - image: docker.io/my-example-frontend-ui:latest
        name: ui
        resources:
          requests:
            ephemeral-storage: 1Gi
        volumeMounts:
        - mountPath: /cache
          name: anEmptyDir
      volumes:
      - emptyDir: {}
        name: anEmptyDir

Messages:
* `[ERROR] Pod.yaml: container "nginx" mounts local ephemeral storage of type emptyDir, but ephemeral storage limits is not declared (EphemeralStorageLimitsAndRequestsDeclared)`
* `[ERROR] Pod.yaml: container "ui" mounts local ephemeral storage of type emptyDir, but ephemeral storage limits is not declared (EphemeralStorageLimitsAndRequestsDeclared)`
<hr>

**Example: Ephemeral Storage used But No Requests Declared**

_Pod.yaml (rendered from test default)_:

    apiVersion: v1
    kind: Pod
    metadata:
      name: emptyDirExample
    spec:
      containers:
      - image: docker.io/sample-panamax-nginx:latest
        name: nginx
        resources:
          limits:
            ephemeral-storage: 6Mi
        volumeMounts:
        - mountPath: /cache
          name: anEmptyDir
      - image: docker.io/my-example-frontend-ui:latest
        name: ui
        resources:
          limits:
            ephemeral-storage: 5Gi
        volumeMounts:
        - mountPath: /cache
          name: anEmptyDir
      volumes:
      - emptyDir: {}
        name: anEmptyDir

Messages:
* `[ERROR] Pod.yaml: container "nginx" mounts local ephemeral storage of type emptyDir, but ephemeral storage requests is not declared (EphemeralStorageLimitsAndRequestsDeclared)`
* `[ERROR] Pod.yaml: container "ui" mounts local ephemeral storage of type emptyDir, but ephemeral storage requests is not declared (EphemeralStorageLimitsAndRequestsDeclared)`
<hr>

**Example: Ephemeral ConfigMap Storage used But No Limits Declared**

_Pod.yaml (rendered from test default)_:

    apiVersion: v1
    kind: Pod
    metadata:
      name: failingConfigMap
    spec:
      containers:
      - image: docker.io/sample-panamax-nginx:latest
        name: nginx
        resources:
          requests:
            ephemeral-storage: 2Mi
            memory: 32Mi
        securityContext:
          readOnlyRootFilesystem: false
        volumeMounts:
        - mountPath: /cache
          name: aConfigMap
      volumes:
      - configMap:
          items:
          - keys: jsonString
            path: data
          name: targetConfigMap
        name: aConfigMap

Messages:
* `[ERROR] Pod.yaml: container "nginx" mounts local ephemeral storage of type configMap, but ephemeral storage limits is not declared (EphemeralStorageLimitsAndRequestsDeclared)`
<hr>

</details>

### Checklist

- [x] Write tests
- [x] Make sure all tests pass
- [x] Update documentation
- [ ] Check RBAC rights for Kubernetes / OpenShift roles
- [x] Try your changes from Pod inside your Kubernetes and OpenShift cluster, not just locally
- [ ] Reference relevant issue(s) and close them after merging
- [ ] Update CHANGELOG.md
- [ ] Supply screenshots for visual changes, such as Grafana dashboards

### CODE REVIEW

**Review by scholzj — CHANGES_REQUESTED** [review]
Sorry, but can you please start by explaining what the PR does? You seem to have copied some part of some documentation, but there is absolutely no explation what the PR does.

File: install/cluster-operator/060-Deployment-strimzi-cluster-operator.yaml [hunk 1]
```diff
@@ -120,5 +120,6 @@ spec:
             requests:
               cpu: 200m
               memory: 384Mi
+              ephemeral-storage: 1Gi
```
Comment (scholzj) [inline install/cluster-operator/060-Deployment-strimzi-cluster-operator.yaml:123]: This file should not change outside of releases. If you want to change something, you need to:
* Do it in the `packaging/...` doirectory
* Change it in the Helm Chart and call `make helm_install` to generate the YAML files

**Review by scholzj — CHANGES_REQUESTED** [review]
Even after reading the code I stillabsolutely don't understand what are you trying to do.

**Comment by Joel-hanson** [comment]
Sorry for the long PR description, I should have give a more clearer message. (have updated the PR description)

What does this PR do?

1. This PR configures requests and limits for ephemeral storage in each product container for pods that utilise it.
2. If the following circumstances are fulfilled, it updates all container pods with ephemeral storage requests and limits:

  - Request - If the volume is either `emptyDir`, `configMap`, `downardAPI`, or `secret`, the ephemeral-storage request is set for the container pod resources in all the circumstances.
  - Limit - For all container pod resources, an ephemeral-storage limit has been set:

    1. If the container security context is non-ReadOnlyRootFilesystem and the `emptyDir` volume doesn't have a sizeLimit.
    2. If the container security context is non-ReadOnlyRootFilesystem and it is one of `configMap,` `downardAPI,` or`secret` then the ephemeral-storage limit is set.

3. Updated Pod template to have the configuration for ephemeral-storage request and ephemeral-storage limit.
4. The test for the above implementation.
5. The default value of the ephemeral storage is set as `1 Gi`


**_Note: The default values for the ephemeral storage is set to `1 Gi` because the teams with the potential need of more than 1GB of local ephemeral storage should consider different options such as persistent storage for its own PVC rather than relying on ephemeral storage._**

File: install/cluster-operator/060-Deployment-strimzi-cluster-operator.yaml (hunk 1 above)
Comment (Joel-hanson) [reply install/cluster-operator/060-Deployment-strimzi-cluster-operator.yaml]: Removed. This was added by mistake.

**Comment by Joel-hanson** [comment]
I apologise for the delay, @scholzj. Please allow me to respond to your points tomorrow .

**Comment by Joel-hanson** [comment]
Closing this PR. I believe that the only PR component that could be required is establishing the ephemeral storage request. We are investigating the ephemeral storage limit portion and will reopen the PR when we later review this adjustment, if necessary.

**Pull request closed without merging**